src/net_processing.cpp

```diff
@@ -459,7 +459,7 @@ struct CNodeState {
 };
 
 // Keeps track of the time (in microseconds) when transactions were requested last time
-limitedmap<uint256, std::chrono::microseconds> g_already_asked_for GUARDED_BY(cs_main)(MAX_INV_SZ);
+limitedmap<GenTxid, std::chrono::microseconds> g_already_asked_for GUARDED_BY(cs_main)(MAX_INV_SZ);
```

Comment:
The reason I'm changing it here is because I have a hard time reasoning about the consistency of the various data structures around tx requesting if we don't. That, plus the fact that I don't really see a way to maintain the same behaviour sanely in #19184 where I'd hope to not duplicate the information any more.

Could this be addressed by instead testing for both WTX+wtxid and TX+wtxid?

Reply:
I haven't yet reviewed #19184, so I'll try to do that to get a better sense of what you're referring to.

> Could this be addressed by instead testing for both WTX+wtxid and TX+wtxid?

I think that would also work, but if that's how we would do all lookups into `g_already_asked_for`, I don't know why we'd bother with that complexity? 

To be clear, my suggested change is fairly minimal code:

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 6cee368ed72..36c890e395b 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -459,7 +459,7 @@ struct CNodeState {
 };
 
 // Keeps track of the time (in microseconds) when transactions were requested last time
-limitedmap<GenTxid, std::chrono::microseconds> g_already_asked_for GUARDED_BY(cs_main)(MAX_INV_SZ);
+limitedmap<uint256, std::chrono::microseconds> g_already_asked_for GUARDED_BY(cs_main)(MAX_INV_SZ);
 
 /** Map maintaining per-node state. */
 static std::map<NodeId, CNodeState> mapNodeState GUARDED_BY(cs_main);
@@ -756,12 +756,12 @@ static void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vec
 
 void EraseTxRequest(const GenTxid& gtxid) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
-    g_already_asked_for.erase(gtxid);
+    g_already_asked_for.erase(gtxid.GetHash());
 }
 
 std::chrono::microseconds GetTxRequestTime(const GenTxid& gtxid) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
-    auto it = g_already_asked_for.find(gtxid);
+    auto it = g_already_asked_for.find(gtxid.GetHash());
     if (it != g_already_asked_for.end()) {
         return it->second;
     }
@@ -770,9 +770,9 @@ std::chrono::microseconds GetTxRequestTime(const GenTxid& gtxid) EXCLUSIVE_LOCKS
 
 void UpdateTxRequestTime(const GenTxid& gtxid, std::chrono::microseconds request_time) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
-    auto it = g_already_asked_for.find(gtxid);
+    auto it = g_already_asked_for.find(gtxid.GetHash());
     if (it == g_already_asked_for.end()) {
-        g_already_asked_for.insert(std::make_pair(gtxid, request_time));
+        g_already_asked_for.insert(std::make_pair(gtxid.GetHash(), request_time));
     } else {
         g_already_asked_for.update(it, request_time);
     }
```

This would capture my understanding of the `g_already_asked_for` data structure: when scheduling transaction requests, we don't care how a given txid/wtxid was requested (via txid or wtxid), just that a getdata for that hash was sent out at a certain time.

Reply:
I see.

It's just a question of whether we treat the tx request data as conceptually "keyed" by hash (and seeing is_wtxid as metadata along with the request), or by (is_wtxid, hash). The fact that we may want to treat a wtxid request for the same hash as a txid request is a sign that it should probably be the first.

I picked the (is_wtxid, hash} based keying first as it seemed easier to reason about, but if consistently applied (treating all requests for the same hash as identical), there doesn't seem to be much difference.

Both are possible to integrate into #19184, with not much complexity difference. But it has to be consistent - we can't be treating them as identical in some cases, and differentiate in others.

I think this would also mean making `m_tx_in_flight` and `m_tx_announced` be hash-keyed (as @ajtwowns' original patch did)?

Reply:
Done.